pgmax/fg/groups.py

```diff
@@ -556,23 +353,48 @@ class FactorGroup:
         variable_group: either a VariableGroup or - if the elements of more than one VariableGroup
             are connected to this FactorGroup - then a CompositeVariableGroup. This holds
             all the variables that are connected to this FactorGroup
-
-    Attributes:
-        _variables_to_factors: maps set of connected variables to the corresponding factors
+        variable_names_for_factors: A list of list of variable names, where each innermost element is the
+            name of a variable in variable_group. Each list within the outer list is taken to contain
+            the names of the variables connected to a factor.
+        log_potentials: Optional array of shape (num_val_configs,) or (num_factors, num_val_configs).
+            If specified, it contains the log of the potential value for every possible configuration.
+            If none, it is assumed the log potential is uniform 0 and such an array is automatically
+            initialized.
     """
 
     variable_group: Union[CompositeVariableGroup, VariableGroup]
-    _variables_to_factors: Mapping[FrozenSet, nodes.Factor] = field(init=False)
+    variable_names_for_factors: Sequence[List]
+    num_factors: int = field(init=False)
+    factor_edges_num_states: np.ndarray = field(init=False)
+    variables_and_num_states: OrderedDict[FrozenSet, nodes.Factor] = field(init=False)
```

Comment:
It seems it's better to change these to `cached_property`s (e.g. `factor_edges_num_states` to be consistent with how it's done for individual factors, and `variables_for_factors` which is the parallel of `variables` in individual factors.

Reply:
I would advocate to not moving to `cached_property` to only loop through the factors once, and save time